File: src/api/routes/compras.py
from fastapi import APIRouter, responses
from pydantic import BaseModel
from datetime import date
from api.services import database_compras, respostas_padrao

router = APIRouter()

class BaseCompra(BaseModel):
    fornecedor_id: int
    data_compra: date | None = None
    previsao_entrega: date | None = None
    status: str = "PENDENTE"

class ItemCompraInput(BaseModel):
    materia_prima_id: int
    quantidade: float
    valor_unitario: float | None = None

@router.get("/")
async def listar_compras():
    resultado = database_compras.listar_compras()
    return responses.JSONResponse(content=resultado, status_code=200)

@router.get("/{id}")
async def listar_compra_id(id: int):
    resultado = database_compras.listar_compra_id(id)
    if resultado is None:
        return respostas_padrao.nao_encontrado
    return responses.JSONResponse(content=resultado, status_code=200)

@router.post("/")
async def criar_compra(body: BaseCompra):
    compra_id = database_compras.criar_compra(
        fornecedor_id=body.fornecedor_id,
        data_compra=body.data_compra,
        previsao_entrega=body.previsao_entrega,
        status=body.status
    )
    return responses.JSONResponse(
        content={"id": compra_id, "status": "Ok"}, status_code=200
    )

@router.put("/{id}")
async def atualizar_compra(id: int, body: BaseCompra):
    resultado = database_compras.editar_compra(
        id=id,
        fornecedor_id=body.fornecedor_id,
        data_compra=str(body.data_compra) if body.data_compra else None,
        previsao_entrega=str(body.previsao_entrega) if body.previsao_entrega else None,
        status=body.status
    )
    return responses.PlainTextResponse(content=resultado, status_code=200)

@router.delete("/{id}")
async def deletar_compra(id: int):
    resultado = database_compras.deletar_compra(id)
    return responses.PlainTextResponse(content=resultado, status_code=200)

@router.post("/{id}/receber")
async def receber_compra(id: int):
    resultado = database_compras.receber_compra(id)
    if resultado != "Ok":
        return responses.JSONResponse(
            content={"erro": resultado}, status_code=400
        )
    return responses.PlainTextResponse(content=resultado, status_code=200)

@router.post("/{id}/cancelar")
async def cancelar_compra(id: int):
    resultado = database_compras.cancelar_compra(id)
    if resultado != "Ok":
        return responses.JSONResponse(
            content={"erro": resultado}, status_code=400
        )
    return responses.PlainTextResponse(content=resultado, status_code=200)

@router.post("/{id}/itens")
async def adicionar_item_compra(id: int, body: ItemCompraInput):
    resultado = database_compras.adicionar_item_compra(
        compra_id=id,
        materia_prima_id=body.materia_prima_id,
        quantidade=body.quantidade,
        valor_unitario=body.valor_unitario
    )
    return responses.PlainTextResponse(content=resultado, status_code=200)

File: src/web/app/js/pages/compras.js
async function renderCompras() {
    const res = await API.listarCompras();
    const data = res.ok ? res.data : [];
    let html = `<div class="d-flex justify-content-between mb-3">
        <p></p>
        <button class="btn btn-primary btn-sm" onclick="compraForm(null)"><i class="bi bi-plus-lg"></i> Nova</button>
    </div>`;
    html += renderTable(
        ["ID", "Fornecedor ID", "Data", "Previsao Entrega", "Recebimento", "Status"],
        data.map(c => [c.id, c.fornecedor_id, c.data_compra, c.previsao_entrega || "-", c.data_recebimento || "-",
            `<span class="badge ${statusBadge(c.status)}">${c.status}</span>`]),
        r => `<button class="btn btn-sm btn-outline-info me-1" onclick="compraForm(${r[0]})"><i class="bi bi-pencil"></i></button>
              <button class="btn btn-sm btn-outline-success me-1" onclick="receberCompra(${r[0]})"><i class="bi bi-check-lg"></i></button>
              <button class="btn btn-sm btn-outline-warning me-1" onclick="cancelarCompra(${r[0]})"><i class="bi bi-x-lg"></i></button>
              <button class="btn btn-sm btn-outline-danger" onclick="compraDelete(${r[0]})"><i class="bi bi-trash"></i></button>`
    );
    document.getElementById("content-body").innerHTML = html;
}

window.compraForm = async function (id) {
    let c = { fornecedor_id: "", data_compra: "", previsao_entrega: "", status: "PENDENTE" };
    if (id) { const r = await API.buscarCompra(id); if (r.ok && r.data) c = r.data; }
    showModal(id ? "Editar Compra" : "Nova Compra",
        formGroup("Fornecedor ID", "f-forn", "number", c.fornecedor_id) +
        formGroup("Data Compra", "f-data", "date", c.data_compra || "") +
        formGroup("Previsao Entrega", "f-prev", "date", c.previsao_entrega || "") +
        selGroup("Status", "f-status", [
            { value: "PENDENTE", label: "Pendente" }, { value: "RECEBIDA", label: "Recebida" }, { value: "CANCELADA", label: "Cancelada" }
        ], c.status),
        async function () {
            const d = { fornecedor_id: Number(val("f-forn")), data_compra: val("f-data") || null, previsao_entrega: val("f-prev") || null, status: val("f-status") };
            if (id) await API.atualizarCompra(id, d); else {
                const cr = await API.criarCompra(d);
                if (cr.ok && cr.data && cr.data.id) {
                    const compraId = cr.data.id;
                    closeModal();
                    await adicionarItensCompra(compraId);
                    return;
                }
            }
            closeModal(); renderCompras();
        }
    );
};

window.adicionarItensCompra = async function (compraId) {
    showModal("Adicionar Item a Compra #" + compraId,
        formGroup("Materia-Prima ID", "f-mp", "number", "") +
        formGroup("Quantidade", "f-qtd", "number", "") +
        formGroup("Valor Unitario", "f-vlr", "number", ""),
        async function () {
            const d = { materia_prima_id: Number(val("f-mp")), quantidade: Number(val("f-qtd")), valor_unitario: val("f-vlr") ? Number(val("f-vlr")) : null };
            await API.adicionarItemCompra(compraId, d);
            if (confirm("Item adicionado! Adicionar mais itens?")) {
                adicionarItensCompra(compraId);
            } else {
                closeModal();
                renderCompras();
            }
        }
    );
};

window.compraDelete = async function (id) { if (confirm("Deletar compra?")) { await API.deletarCompra(id); renderCompras(); } };
window.receberCompra = async function (id) { if (confirm("Receber compra?")) { await API.receberCompra(id); renderCompras(); } };
window.cancelarCompra = async function (id) { if (confirm("Cancelar compra?")) { await API.cancelarCompra(id); renderCompras(); } };
